### PR TITLE
[SPARK-46792][PYTHON][CONNECT] Refactor ChannelBuilder into DefaultChannelBuilder and ChannelBuilder

### PR DESCRIPTION
**Pull request by cdkrot**
### What changes were proposed in this pull request?

Refactor ChannelBuilder into ChannelBuilder and DefaultChannelBuilder. ChannelBuilder is a abstract class with certain base functionality other possible channel builders should inherit. DefaultChannelBuilder is the current implementation of ChannelBuilder.

### Why are the changes needed?

The concept of ChannelBuilder was introduced specifically so that different implementations of Channel Builder would exist. However, the current version makes it hard to actually implement custom functionality, because the interface of channel builder and it's standard implementation are tied so closely.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Existing tests; new e2e test.

### Was this patch authored or co-authored using generative AI tooling?
Nope :)

### CODE REVIEW

File: python/pyspark/sql/connect/client/core.py [hunk 1]
```diff
@@ -450,18 +486,18 @@ def keys(self) -> List[str]:
...
-        tree_string: Optional[str],
-        is_local: Optional[bool],
-        is_streaming: Optional[bool],
-        input_files: Optional[List[str]],
-        spark_version: Optional[str],
-        parsed: Optional[DataType],
-        is_same_semantics: Optional[bool],
-        semantic_hash: Optional[int],
-        storage_level: Optional[StorageLevel],
+            self,
```
Comment (HyukjinKwon) [inline python/pyspark/sql/connect/client/core.py:489]: Hm, we won;t need formatting here. Was it by python auto formatter?

File: python/pyspark/sql/connect/client/core.py [hunk 2]
```diff
@@ -15,7 +15,8 @@
 # limitations under the License.
 #
 __all__ = [
-    "ChannelBuilder",
+    "DefaultChannelBuilder",
+    "DefaultChannelBuilder",
```
Comment (HyukjinKwon) [inline python/pyspark/sql/connect/client/core.py:19]: remove one?

File: python/pyspark/sql/connect/client/core.py [hunk 3]
```diff
@@ -100,26 +101,183 @@ class ChannelBuilder:
     This is a helper class that is used to create a GRPC channel based on the given
     connection string per the documentation of Spark Connect.
 
+    The standard implementation is in DefaultChannelBuilder.
```
Comment (HyukjinKwon) [inline python/pyspark/sql/connect/client/core.py:104]:
```suggestion
    The standard implementation is in :class:`DefaultChannelBuilder`.
```

**Review by HyukjinKwon — COMMENTED** [review]
Seems fine but should probably cleanup and remove unnecessary changes

File: python/pyspark/sql/connect/client/core.py (hunk 2 above)
Comment (cdkrot) [reply python/pyspark/sql/connect/client/core.py:19]: Should have been "ChannelBuilder", "DefaultChannelBuilder", thanks!

File: python/pyspark/sql/connect/client/core.py (hunk 1 above)
Comment (cdkrot) [reply python/pyspark/sql/connect/client/core.py:489]: Used autoformatter and it get back to reason, so probably IDE did. I'm changing it back, sorry :)

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Let's fix up the linter failure

**Comment by cdkrot** [comment]
Thanks! Tests green now

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**